RemoveImport should look through supertypes for Kotlin

jevanlingen opened the pull request:
## What's changed?
The remove import recipe takes imports of super types into account.

## What's your motivation?
Kotlin can import a static java member of a parent class by importing the child class. The recipe did not take this feature into account. There's also the option of using imports of companion objects with an interface. These kind of setup was not supported as well.

## Any additional context
At the moment I could not find a proper way to add Java source code as dependency of the Kotlin sources, so I added a `@Disabled` test. If there's a way to do this, it should be used.

### Checklist
- [x] I've added unit tests to cover both positive and negative cases
- [x] I've read and applied the [recipe conventions and best practices](https://docs.openrewrite.org/authoring-recipes/recipe-conventions-and-best-practices)
- [x] I've used the IntelliJ IDEA auto-formatter on affected files

Laurens-W commented on rewrite-java/src/main/java/org/openrewrite/java/RemoveImport.java:
```diff
@@ -58,48 +58,85 @@ public RemoveImport(String type, boolean force) {
+                        // For Kotlin, the owning class interfaces with methods can be used without actually importing those interfaces directly...
+                        JavaType.Class owningClass = TypeUtils.asClass(fq.getOwningClass());
+                        if (owningClass != null) {
+                            Queue<JavaType.FullyQualified> toVisit = new LinkedList<>(owningClass.getInterfaces());
+                            Set<JavaType.FullyQualified> visited = new HashSet<>();
+                            while (!toVisit.isEmpty()) {
+                                JavaType.FullyQualified current = toVisit.poll();
+                                if (!visited.add(current)) {
+                                    continue;
+                                }
+                                toVisit.addAll(current.getInterfaces());
+                            }
```
Not sure if we'd prefer nested forloops here instead of a Queue

Laurens-W replied:
I must've misread the code initially, thinking it could be solved with a nested forloop. Your arguments for keeping it like this are strong and valid :)

timtebeek commented:
### Alternative 1 — extend the `ImportService` SPI (instead of Kotlin-aware logic in `rewrite-java`)

`maybeAddImport` and `maybeRemoveImport` are dispatched differently in `JavaVisitor`:

- `maybeAddImport` → `service(ImportService.class).addImportVisitor(...)` — **polymorphic**. `KotlinImportService extends ImportService` already overrides it to return the Kotlin-specific `org.openrewrite.kotlin.AddImport`.
- `maybeRemoveImport` → `new RemoveImport<>(...)` — **hardcoded** in the Java module. There is no language seam, which is *why* this PR injects Kotlin awareness (path sniffing, `"Kt"`/`"Kts"` facade handling, supertype walking) into the shared `org.openrewrite.java.RemoveImport`.

So `AddImport` already has exactly the separation `RemoveImport` is missing. The alternative:

1. Add `removeImportVisitor(...)` to `ImportService`.
2. Route `JavaVisitor#maybeRemoveImport` through `service(ImportService.class)` instead of `new RemoveImport<>(...)`.
3. Override `removeImportVisitor` in `KotlinImportService` to return a new `org.openrewrite.kotlin.RemoveImport`, mirroring `org.openrewrite.kotlin.AddImport`.

**Pros:** mirrors `AddImport` exactly; removes all `isKotlin` source-path sniffing from `rewrite-java`; uses the type system rather than a filename heuristic; Kotlin-specific logic (facade convention, supertype walk) lives in `rewrite-kotlin`.

**Cons:** more moving parts (new SPI method + new class); `RemoveImport` carries a lot of shared logic, so the Kotlin subclass would either duplicate it or require extracting shared helpers into a common base.

timtebeek commented:
### Alternative 3 — fix it at type attribution in the Kotlin parser

The root cause this PR works around is how Kotlin attributes the relevant types:

- top-level Kotlin functions get a `null` `owningClass` and a `<File>Kt` facade `declaringType`, even though they are imported by their package-qualified name (`org.example.one`);
- members inherited by a companion object (e.g. from an implemented interface) aren't matched against the import without walking supertypes.

If `KotlinTypeMapping` attributed these so that the **existing** `RemoveImport` logic already matched (e.g. the top-level function carried its package-qualified name, inherited members resolved through the supertype), then no change to `RemoveImport` would be needed at all.

**Pros:** fixes the data rather than every consumer; the most "correct" location — any recipe relying on these types benefits, not just `RemoveImport`.

**Cons:** highest blast radius — touches type attribution used everywhere, so regression risk is much larger; and the facade↔import-name mapping is arguably a presentation concern that may not belong on the type itself.

timtebeek committed:
RemoveImport: look through supertypes and top-level functions in Kotlin

Rebased onto current main (flattened from the original branch's polish/merge
history) and updated for current Kotlin type attribution:

- Look through Kotlin supertypes so members inherited by a companion object
  (e.g. from an implemented interface) keep their import.
- Map top-level Kotlin functions from their '<File>Kt' facade class back to
  their package-qualified import name. Top-level functions now have a null
  owningClass, which the original facade-walk relied on.
- Guard type.substring(type.lastIndexOf('.')) against types without a package
  to avoid StringIndexOutOfBoundsException (RemoveImportTest.doesNotCrashWhenTypeHasNoPackage).

timtebeek commented:
I've pushed up the alternative that does not have any Kotlin specific handling built into this recipe that operates across a growing number of languages. 
- https://github.com/openrewrite/rewrite/pull/8159

Thanks for helping explore the problem space!
